### PR TITLE
add crash-report capabilities

### PR DESCRIPTION
**Pull request by umbynos**
**Please check if the PR fulfills these requirements**

- [X] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-create-agent/pulls)
      before creating one)

* **What kind of change does this PR introduce?**

feature
- **What is the current behavior?**

Sometimes the agent crashes without apparent reason
* **What is the new behavior?**

A folder containing crash-reports is generated in that same folder containing the executable (e.g. in Linux it's in `~/ArduinoCreateAgent-<version>/logs`) when the agent crashes. It should allow users to better report issues, and developers to better understand and fix bugs.
The file created inside the folder is named `crashreport_yyyyMMddHHmmss.log` and should contain something like:
```
stacktrace from panic: 
goroutine 1 [running, locked to thread]:
runtime/debug.Stack(0x0, 0x0, 0x0)
	/home/linuxbrew/.linuxbrew/Cellar/go/1.15.3/libexec/src/runtime/debug/stack.go:24 +0xa5
main.panicToFile()
	/home/umberto/Nextcloud/8tb/Lavoro/arduino-create-agent/main.go:165 +0x22d
panic(0xa09400, 0xb0ebb0)
	/home/linuxbrew/.linuxbrew/Cellar/go/1.15.3/libexec/src/runtime/panic.go:975 +0x4c6
main.main()
	/home/umberto/Nextcloud/8tb/Lavoro/arduino-create-agent/main.go:118 +0x98
```
- **Does this PR introduce a breaking change?**

no
* **Other information**:

- [X] add crash-report generation
- [X] allow users to enable/disable the functionality
- [x] allow users to remove all the generated logs

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -148,6 +151,24 @@ func main() {
 	Systray.Start()
 }
 
+func panicToFile() {
+	if r := recover(); r != nil {
+		fileName := "crashreport_" + time.Now().Format("20060102150405") + ".txt"
+		currDir, err := osext.ExecutableFolder()
+		if err != nil {
+			panic(err)
```
Comment (umbynos) [inline main.go:159]: I don't know if it's correct to panic inside a `defer` function. Maybe it's better to log the error. Returning the error is not possible.

File: main.go [hunk 2]
```diff
@@ -111,6 +111,9 @@ func main() {
 		os.Exit(0)
 	}
 
+	// function used to save panic to a file (WIP)
+	defer panicToFile()
+
```
Comment (umbynos) [inline main.go:116]: To try the feature I added
`panic("test")` after the `defer` and the file generated contains this stacktrace:
```
stacktrace from panic: 
goroutine 1 [running, locked to thread]:
runtime/debug.Stack(0x0, 0x0, 0x0)
	/home/linuxbrew/.linuxbrew/Cellar/go/1.15.3/libexec/src/runtime/debug/stack.go:24 +0xa5
main.panicToFile()
	/home/umberto/Nextcloud/8tb/Lavoro/arduino-create-agent/main.go:165 +0x22d
panic(0xa09400, 0xb0ebb0)
	/home/linuxbrew/.linuxbrew/Cellar/go/1.15.3/libexec/src/runtime/panic.go:975 +0x4c6
main.main()
	/home/umberto/Nextcloud/8tb/Lavoro/arduino-create-agent/main.go:117 +0x98
```

And this is stderr:
```
panic: test [recovered]
	panic: test

goroutine 1 [running, locked to thread]:
main.panicToFile()
	/home/umberto/Nextcloud/8tb/Lavoro/arduino-create-agent/main.go:170 +0x49b
panic(0xa09400, 0xb0ebb0)
	/home/linuxbrew/.linuxbrew/Cellar/go/1.15.3/libexec/src/runtime/panic.go:975 +0x4c6
main.main()
	/home/umberto/Nextcloud/8tb/Lavoro/arduino-create-agent/main.go:117 +0x98
```

**Comment by umbynos** [comment]
Problems still to be solved:
1. if the panic happens in the [loop() function ](https://github.com/arduino/arduino-create-agent/blob/e53983708775f4c4ca8cde64dc295486cece913f/main.go#L184) the panic does not get stopped because `loop()` is run in a different goroutine.
> The process continues up the stack until all functions in the **current**  goroutine have returned, at which point the program crashes
https://blog.golang.org/defer-panic-and-recover

2. Handling of the old crash-reports still missing (which is the best way to do it?)

**Comment by umbynos** [comment]
I've changed the approach: now everything happening on stderr is redirected to the crash-report saved locally. I've followed [this issue](https://stackoverflow.com/questions/34772012/capturing-panic-in-golang) There's still work to do regarding stdout and stderr. Now every output is redirected to the file (maybe because stdout and stderr are treated the same way(?))
Another possible problem concerns crash-report handling (how not to spam too many files)

**Comment by umbynos** [comment]
Example of crash-report generated on windows.
[crashreport_20201201165111.txt](https://github.com/arduino/arduino-create-agent/files/5624201/crashreport_20201201165111.txt)

**Comment by zmoog** [comment]
* Add a command in the systray to clean the logs
* rename .txt > .log

**Comment by umbynos** [comment]
Now the user should be able to optionally enable crash-report generation changing `crashreport = true` in `config.ini`
By default the functionally is disabled (`crashreport = false`).

**Comment by umbynos** [comment]
image
Added option in the trayicon's menu to remove the crashreports
image
The newly added item is disables if the folder `logs` is not present.
Also logging have been added to the code added

**Review by zmoog — APPROVED** [review]
LGTM ❤️

**Pull request merged**